worker/src/ocall_bridge/attestation_ocall_impl.rs

```diff
@@ -0,0 +1,180 @@
+/*
+    Copyright 2019 Supercomputing Systems AG
+    Copyright (C) 2017-2019 Baidu, Inc. All Rights Reserved.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+        http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+*/
+
+use crate::ocall_bridge::bridge_api::{
+    OCallBridgeError, OCallBridgeResult, RemoteAttestationOCall,
+};
+use frame_support::ensure;
+use log::*;
+use sgx_types::*;
+use std::net::{SocketAddr, TcpStream};
+use std::os::unix::io::IntoRawFd;
+use std::ptr;
+
+pub struct RemoteAttestationOCallImpl {
+    // TODO as a member here we need the e-call API trait, so we can use it instead of making the e-call directly
+}
+
+impl RemoteAttestationOCall for RemoteAttestationOCallImpl {
+    fn init_quote(&self) -> OCallBridgeResult<(sgx_target_info_t, sgx_epid_group_id_t)> {
+        // TODO this translation to unsafe C-API should be moved to the EnclaveApi / ECall API
+        let mut ti: sgx_target_info_t = sgx_target_info_t::default();
+        let mut eg: sgx_epid_group_id_t = sgx_epid_group_id_t::default();
+
+        let ret_status = unsafe {
+            sgx_init_quote(
+                &mut ti as *mut sgx_target_info_t,
+                &mut eg as *mut sgx_epid_group_id_t,
+            )
+        };
+
+        ensure!(
+            ret_status == sgx_status_t::SGX_SUCCESS,
+            OCallBridgeError::InitQuote(ret_status)
+        );
+
+        Ok((ti, eg))
+    }
+
+    fn get_ias_socket(&self) -> OCallBridgeResult<i32> {
+        let port = 443;
+        let hostname = "api.trustedservices.intel.com";
+
+        let addr = lookup_ipv4(hostname, port).map_err(OCallBridgeError::GetIasSocket)?;
+
+        let sock = TcpStream::connect(&addr).map_err(|_| {
+            OCallBridgeError::GetIasSocket("[-] Connect tls server failed!".to_string())
```

Comment:
I only see this error here. Is there a reason to use `OcallBridgeError::GetIasSocker(String)` over `OcallBridgeError::GetIasSocker(TcpError)` (or whatever error they use in this crate)?

Reply:
The line just above that, the call to `lookup_ipv4`, has a different error type and is not related to `TcpStream`. So I don't think it makes much sense to use `TcpError` in my own error definition?

Reply:
Wow, somehow my ctrl + f did not see this. Then I think it's perfectly fine.